pallets/capacity/src/lib.rs

```diff
@@ -0,0 +1,658 @@
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Tokens have been staked to the Frequency network.
+		Staked {
+			/// The token account that staked tokens to the network.
+			account: T::AccountId,
+			/// The MSA that a token account targeted to receive Capacity based on this staking amount.
+			target: MessageSourceId,
+			/// An amount that was staked.
+			amount: BalanceOf<T>,
+			/// The Capacity amount issued to the target as a result of the stake.
+			capacity: BalanceOf<T>,
+		},
+		/// Unsstaked token that has thawed was unlocked for the given account
+		StakeWithdrawn {
+			/// the account that withdrew its stake
+			account: T::AccountId,
+			/// the total amount withdrawn, i.e. put back into free balance.
+			amount: BalanceOf<T>,
+		},
+		/// A token account has unstaked the Frequency network.
+		UnStaked {
+			/// The token account that staked tokens to the network.
+			account: T::AccountId,
+			/// The MSA that a token account targeted to receive Capacity to unstake from.
+			target: MessageSourceId,
+			/// An amount that was unstaked.
+			amount: BalanceOf<T>,
+			/// The Capacity amount that was reduced from a target.
+			capacity: BalanceOf<T>,
+		},
+		/// The Capacity epoch length was changed.
+		EpochLengthUpdated {
+			/// The new length of an epoch in blocks.
+			blocks: T::BlockNumber,
+		},
+		/// Capacity has been withdrawn from a MessageSourceId.
+		CapacityWithdrawn {
+			/// The MSA from which Capacity has been withdrawn.
+			msa_id: MessageSourceId,
+			/// The amount of Capacity withdrawn from MSA.
+			amount: BalanceOf<T>,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Staker attempted to stake to an invalid staking target.
+		InvalidTarget,
+		/// Capacity is not available for the given MSA.
+		InsufficientBalance,
+		/// Staker is attempting to stake an amount below the minimum amount.
+		InsufficientStakingAmount,
+		/// Staker is attempting to stake a zero amount.
+		ZeroAmountNotAllowed,
+		/// Origin has no Staking Account
+		NotAStakingAccount,
+		/// No staked value is available for withdrawal; either nothing is being unstaked,
+		/// or nothing has passed the thaw period.
+		NoUnstakedTokensAvailable,
+		/// Unstaking amount should be greater than zero.
+		UnstakedAmountIsZero,
+		/// Amount to unstake is greater than the amount staked.
+		AmountToUnstakeExceedsAmountStaked,
+		/// Attempting to unstake from a target that has not been staked to.
+		StakingAccountNotFound,
+		/// Attempting to get a staker / target relationship that does not exist.
+		StakerTargetRelationshipNotFound,
+		/// Attempting to get the target's capacity that does not exist.
+		TargetCapacityNotFound,
+		/// Staker reached the limit number for the allowed amount of unlocking chunks.
+		MaxUnlockingChunksExceeded,
+		/// Increase Capacity increase exceeds the total available Capacity for target.
+		IncreaseExceedsAvailable,
```

Comment:
I don't see `IncreaseExceedsAvailable` being used anywhere.

Reply:
Hmm, this is a good one. Let me ask tomorrow about how we should address this.